File: windows/src/TraceCommons.Interop/ReadGate.cs
using System;

namespace TraceCommons.Interop;

/// <summary>
/// What the preview sheet requires, and what it says, at the moment of
/// consent.
///
/// <para>
/// <b>This used to be a read gate.</b> Contribute waited on three things: a
/// pinned preview, "Exactly what would be sent" having actually been on
/// screen, and an acknowledgement ticked by hand. Two of them are gone.
/// </para>
/// <para>
/// The checkbox was removed as friction. The transcript-shown condition
/// went with it, for a reason worth recording: a queue row's Submit
/// approves the same session with no preview opened at all, so the gate
/// never stood between anybody and a blind approval -- it only charged a
/// click to the one contributor who chose to look. A control that taxes
/// the careful path and stops nobody is not a safety property.
/// </para>
/// <para>
/// <b>What did not go is the claim.</b> The gate statement carries both
/// halves of what the checkbox made a contributor assert -- scrubbing is
/// pattern-based and may have missed something, and nothing here can tell
/// whether anyone read anything -- and the sheet prints it above Contribute
/// where the tick used to be asked for.
/// </para>
/// <para>
/// And <see cref="HasPinnedPreview"/> stayed. That was never friction: an
/// approval binds to the envelope a preview pinned, and a preview that
/// failed or was built without an enrollment pinned nothing. Nothing here
/// is persisted and nothing starts set, so every sheet starts from zero and
/// "the bytes the contributor was shown" means this session's bytes.
/// </para>
/// <para>
/// The sentences moved. They are composed once in
/// <c>crates/trace-commons-contributor/src/consent_copy.rs</c> and read here
/// through <see cref="ConsentSurface"/>; what is left in this class is the
/// rule, which is testable on a machine that cannot build WinUI and cannot
/// load the cdylib. The Rust test that used to open this file and grep it
/// for the claim is gone with them: three shells reading one constant is a
/// stronger thing than three shells grepping each other.
/// </para>
/// </summary>
public sealed class ReadGate
{
    /// <summary>Raised whenever the one condition changes.</summary>
    public event Action? Changed;

    /// <summary>
    /// Whether a real preview is loaded and pinned. Set from
    /// <see cref="PreviewSummary.Enrolled"/>; a failed or unenrolled preview
    /// leaves it false.
    /// </summary>
    public bool HasPinnedPreview { get; private set; }

    /// <summary>The one question the sheet asks this object.</summary>
    public bool CanContribute => HasPinnedPreview;

    /// <summary>
    /// Records that a pinned preview is available. A summary that failed to
    /// parse, or one built without an enrollment, must pass false.
    /// </summary>
    public void SetPinnedPreview(bool pinned)
    {
        if (HasPinnedPreview == pinned)
        {
            return;
        }

        HasPinnedPreview = pinned;
        Changed?.Invoke();
    }

    /// <summary>
    /// Clears the pin.
    ///
    /// Called when the sheet moves to a different session. A pin that
    /// carried over would let the second session be approved against the
    /// first one's envelope.
    /// </summary>
    public void Reset() => SetPinnedPreview(false);
}

File: macos/Sources/TCShellCore/ReadGate.swift
import Foundation

/// What the preview sheet requires at the moment of consent.
///
/// ## What this used to be
///
/// `Contribute` used to wait on three things: a loaded preview, the
/// "Exactly what would be sent" tab having been on screen, and an
/// acknowledgement checkbox ticked by hand. Two of them are gone.
///
/// The checkbox was removed as friction. The transcript-tab condition went
/// with it, for a reason worth writing down: a queue row's `Submit`
/// approves the same session without opening the preview at all, so the
/// gate never stood between anybody and a blind approval. The only person
/// it ever charged was the one who chose to look, which is the opposite of
/// what it was for.
///
/// ## What did not go
///
/// The claim. It carries both halves of what the checkbox made a
/// contributor assert -- scrubbing is pattern-based and may have missed
/// something, and nothing here can tell whether anyone read anything -- and
/// the sheet prints it above `Contribute` where the tick used to be asked
/// for. Dropping the friction is a product decision; dropping the sentence
/// would be the app quietly claiming less about redaction than it knows.
///
/// And the pin: an approval still has to cover a preview that actually
/// loaded and carries an enrollment. That is not friction, it is the thing
/// the approval binds to.
///
/// ## Why it lives in TCShellCore
///
/// The same reason `SubmitToast` does. The app target links the FFI dylib,
/// so nothing in it is reachable from `swift test`; a rule that three
/// shells have to agree on needs somewhere it can actually be asserted.
///
/// The sentences moved. They are composed once in
/// `crates/trace-commons-contributor/src/consent_copy.rs` and read here
/// through `TCBridge.TCConsentCopy`; what is left in this enum is the rule,
/// which is testable in a target that links no dylib. The Rust test that
/// used to open this file and grep it for the claim is gone with them.
public enum ReadGate {
    /// The one question the sheet asks.
    ///
    /// A single condition, deliberately. It is stated as a function rather
    /// than inlined into the view so the rule has somewhere to be tested
    /// with values; the view has no testable seam at all.
    public static func canContribute(hasPinnedPreview: Bool) -> Bool {
        hasPinnedPreview
    }
}
